schema/ridl/parser.go

```diff
@@ -0,0 +1,587 @@
+	"fmt"
+	"log"
+)
+
+var (
+	errUnexpectedEOF = errors.New(`unexpected EOF`)
+)
+
+type methodArgument struct {
+	left   *token
+	right  *token
+	stream bool
+}
+
+type method struct {
+	name    *token
+	inputs  []*methodArgument
+	outputs []*methodArgument
+}
+
+type service struct {
+	name    *token
+	methods []*method
+}
+
+type definition struct {
+	left     *token
+	right    *token
+	optional bool
+
+	meta []*definition
+}
+
+type enum struct {
+	name     *token
+	enumType *token
+	values   []*definition
+}
+
+type message struct {
+	name   *token
+	fields []*definition
+}
+
+func (d *definition) value() string {
+	return d.right.val
+}
+
+type tree struct {
+	definitions map[string]*definition
+
+	imports  []*token
+	enums    []*enum
+	messages []*message
+	services []*service
+}
+
+var (
+	eofToken = &token{tt: tokenEOF}
+)
+
+type parseState func(*parser) parseState
+
+type parser struct {
+	tokens []token
+	length int
+	pos    int
+
+	lastErr error
+
+	tree tree
+}
+
+func newParser(input string) (*parser, error) {
+	tokens, err := tokenize(input)
+	if err != nil {
+		return nil, err
+	}
+	p := &parser{
+		tokens: tokens,
+		length: len(tokens),
+		tree: tree{
+			definitions: make(map[string]*definition),
+			imports:     []*token{},
+			enums:       []*enum{},
+		},
+	}
+	return p, nil
+}
+
+func (p *parser) run() error {
+	for state := parseStateStart; state != nil; {
+		state = state(p)
+	}
+	if p.lastErr != nil {
+		return p.lastErr
+	}
+	return nil
+}
+
+func (p *parser) continueUntilEOL() error {
+	for {
+		tok := p.cursor()
+		log.Printf("drop: %v", tok)
+
+		if tok.tt == tokenNewLine {
+			// EOL
+			return nil
+		}
+
+		if !p.next() {
+			// EOF
+			return nil
+		}
+	}
+
+	panic("reached")
+}
+
+func (p *parser) expectCommentOrNewLine() error {
+	for {
+
+		tok := p.cursor()
+		log.Printf("discard?: %q", tok.val)
+
+		switch tok.tt {
+		case tokenNewLine:
+			// got new line, continue
+		case tokenExtra:
+			// unknown char
+		case tokenEOF:
+			// got EOF
+			return nil
+		case tokenHash:
+			if !p.next() {
+				return nil
+			}
+			return p.continueUntilEOL()
+		default:
+			// another kind of token, stop
+			log.Printf("discard? -> nope")
+			return nil
+		}
+		if !p.next() {
+			return nil
+		}
+	}
+	return nil
+}
+
+func (p *parser) expectNext(tt tokenType) error {
+	if !p.next() {
+		return errors.New(`unexpected EOF`)
+	}
+
+	tok := p.cursor()
+	if tok.tt != tt {
+		return fmt.Errorf(`expecting %v, got %v`, tt, tok)
+	}
+	return nil
+}
+
+func (p *parser) stateError(err error) parseState {
+	p.lastErr = fmt.Errorf("error: %v, near: %v \n", err, p.cursor())
+	return nil
+}
+
+func (p *parser) next() bool {
+	if p.pos >= p.length {
+		log.Printf("EOF")
+		return false
+	}
+	p.pos = p.pos + 1
+	return true
+}
+
+func (p *parser) cursor() *token {
+	if p.pos >= p.length {
+		return eofToken
+	}
+	return &p.tokens[p.pos]
+}
+
+func parseStateNewLine(p *parser) parseState {
+	for p.next(); p.cursor().tt == tokenNewLine; {
```

Comment:
there is a bug here, should be `for ; p.cursor().tt == tokenNewLine; p.next() {`